### ISSUE TITLE
typedef-whitespace call-signature false positive with multi-line args list and indentation

### ISSUE THREAD

**Issue by kourge** [palantir/tslint#955]
With typedef-whitespace call-signature set to `nospace`, this is recognized as a warning:
```typescript
function f() : any {}
```
And this passes:
```typescript
function f(): any {}
```
When arguments are spread across multiple lines, indentation produces a false positive, e.g. this works:
```typescript
function f(
  a: any,
  b: any
): any {}
```
Whereas this is identified as violating `nospace`:
```typescript
  function f(
    a: any,
    b: any
  ): any {}
```
Notice that the only difference is in indentation.

**Comment by adidahiya** [comment]
Looks like a bug, thanks for the report @kourge

**Comment by pe8ter** [comment]
I found a related false negative. This passes linting but should actually fail under the `nospace` rule:

```
foo( ) : void {}
```

No multiple lines or tabs involved. What both these failure cases have in common is that the last character in the parameter list is not followed immediately by the closing parenthesis. The AST returned by the TypeScript module doesn't provide enough explicit information to locate the closing parenthesis with a quick one-line fix, but a helper function could fairly quickly be created to find it. I'll keep poking at it.

**Comment by pe8ter** [comment]
To be clearer, the linter incorrectly assumes that a closing parenthesis follows the last character of the parameter list (see lines that contain `node.parameters.end + 1` in _typedefWhitespaceRule.ts_). The position of the closing parenthesis could be far beyond the last character in the parameter list, including after spaces and comments, but that information isn't available in the AST that TypeScript outputs. The only way to truly find the parenthesis is to use the end position of the parameter list `node.parameters.end` and the start of the return type `node.type.pos` combined with a scanner to extract the character run that contains a parenthesis somewhere in it and a RegEx to test for the existence of an offending space.

**Comment by JKillian** [comment]
@pe8ter I'm thinking the AST might be able to help us here infact. I stuck the code below into Atom and used the Ast Full atom-typescript command. The key difference between ast/ast full in atom-typescript is that the first uses `ts.forEachChild` which only iterates over important child nodes, whereas the latter iterates over `node.getChildren` which contains all children. [As explained here](https://basarat.gitbooks.io/typescript/content/docs/compiler/ast-tip-children.html). My hypothesis is that we can just find the colon token node (highlighted in red below) and work off of it.
```
function f(
  a: any,
  b: any
): any {};
```


![image](https://cloud.githubusercontent.com/assets/1400151/12829564/df1e9a0c-cb58-11e5-8cc2-475e4b6bd0bd.png)

**Comment by pe8ter** [comment]
Awesome! I did not know you could get a deeper tree. Will still keep poking. Thanks!
